REGRESSION (iOS 26): Smart Reply text is black when in dark mode

pxlcoder opened the pull request:
#### 4a42bf61dfe36584d85b7a1ac6265ad32f2835bc
<pre>
REGRESSION (iOS 26): Smart Reply text is black when in dark mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=297602">https://bugs.webkit.org/show_bug.cgi?id=297602</a>
<a href="https://rdar.apple.com/157514394">rdar://157514394</a>

Reviewed by Abrar Rahman Protyasha and Richard Robinson.

In iOS 26, Mail removed their use of `-apple-color-filter: apple-invert-lightness()`
in cases where the content explicitly supports dark mode. This change broke
smart replies, which are currently always inserted with a black text color
specified in inline style. When the color filter was present, the black text
would become white.

A black text color is specified in inline style because the attributed string
provided by Writing Tools does not have a text color, and markup sanitization
introduces inline style. The context attributed string given to Writing Tools
also does not contain text color information. However, even after adding color
information at context retrieval time, Writing Tools does not preserve the color.
The lack of color preservation is specific to Smart Reply and not other compositions.

Consequently, fix the issue by updating the attributed string *returned* from
Writing Tools use the body&apos;s text color. When the markup is inserted into the
document, the inline style will be removed due to deduplication.

* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::attributedStringApplyingBodyTextColorIfNecessary):

Don&apos;t apply a text color if one is specified.

(WebCore::WritingToolsController::replaceContentsOfRangeInSession):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, SmartRepliesTextColor)):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView forceLightMode]):

Canonical link: <a href="https://commits.webkit.org/298968@main">https://commits.webkit.org/298968@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5cf9dd8fa9c02548fa8e8a87a1258198174fae32

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117244 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36914 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27548 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123342 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69229 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e819c485-1ae5-4e3a-8306-85d2df5a67ce) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119119 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37610 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45501 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88975 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/43650 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/aabc1033-9879-4498-b023-ce62e6c6c9f0) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120275 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29933 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105138 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69475 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/73716dda-663e-408b-8e0a-6c5f09fbd639) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28993 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23252 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67016 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99331 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23434 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126464 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44141 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33157 "Found 1 new test failure: ipc/send-gradient.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97645 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44497 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101374 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97439 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42791 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20730 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40491 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18725 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44014 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49673 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43470 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46815 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45166 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -3530,6 +3530,67 @@ InstanceMethodSwizzler swizzler(PAL::getWTWritingToolsClass(), @selector(showToo
+    __block bool done = false;
+    [[webView writingToolsDelegate] willBeginWritingToolsSession:session.get() requestContexts:^(NSArray<WTContext *> *contexts) {
+        [[webView writingToolsDelegate] writingToolsSession:session.get() didReceiveAction:WTActionCompositionRestart];
+        TestWebKitAPI::Util::runFor(0.1_s);
```
does a (stable) presentation update not work?

pxlcoder replied:
It does! Updated.

pxlcoder committed:
REGRESSION (iOS 26): Smart Reply text is black when in dark mode
https://bugs.webkit.org/show_bug.cgi?id=297602
rdar://157514394

Reviewed by Abrar Rahman Protyasha and Richard Robinson.

In iOS 26, Mail removed their use of `-apple-color-filter: apple-invert-lightness()`
in cases where the content explicitly supports dark mode. This change broke
smart replies, which are currently always inserted with a black text color
specified in inline style. When the color filter was present, the black text
would become white.

A black text color is specified in inline style because the attributed string
provided by Writing Tools does not have a text color, and markup sanitization
introduces inline style. The context attributed string given to Writing Tools
also does not contain text color information. However, even after adding color
information at context retrieval time, Writing Tools does not preserve the color.
The lack of color preservation is specific to Smart Reply and not other compositions.

Consequently, fix the issue by updating the attributed string *returned* from
Writing Tools use the body's text color. When the markup is inserted into the
document, the inline style will be removed due to deduplication.

* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::attributedStringApplyingBodyTextColorIfNecessary):

Don't apply a text color if one is specified.

(WebCore::WritingToolsController::replaceContentsOfRangeInSession):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, SmartRepliesTextColor)):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView forceLightMode]):

Canonical link: https://commits.webkit.org/298968@main
